frontend/packages/knative-plugin/src/utils/fetch-dynamic-eventsources-utils.ts

```diff
@@ -0,0 +1,93 @@
+import * as _ from 'lodash';
+import { coFetch } from '@console/internal/co-fetch';
+import { K8sKind, kindToAbbr } from '@console/internal/module/k8s';
+import { CustomResourceDefinitionModel } from '@console/internal/models';
+import { chart_color_red_300 as knativeEventingColor } from '@patternfly/react-tokens';
+import {
+  EventSourceCronJobModel,
+  EventSourceContainerModel,
+  EventSourceApiServerModel,
+  EventSourceSinkBindingModel,
+  EventSourceKafkaModel,
+  EventSourceCamelModel,
+} from '../models';
+
+interface EventSourceData {
+  models: K8sKind[];
+}
+
+const defaultEventSourceKind = [
+  EventSourceApiServerModel,
+  EventSourceCamelModel,
+  EventSourceContainerModel,
+  EventSourceCronJobModel,
+  EventSourceKafkaModel,
+  EventSourceSinkBindingModel,
+];
+
+export const eventSourceData: EventSourceData = {
+  models: defaultEventSourceKind,
+};
+
+export const getEventSourcesModel = async () => {
+  let eventSourceModelList: K8sKind[] = [];
+  const url = `api/kubernetes/apis/${CustomResourceDefinitionModel.apiGroup}/${
+    CustomResourceDefinitionModel.apiVersion
+  }/${CustomResourceDefinitionModel.plural}?limit=250&labelSelector=${encodeURIComponent(
+    'duck.knative.dev/source=true',
+  )}`;
+  try {
+    const res = await coFetch(url);
+    const resolvedRes = await res.json();
+    eventSourceModelList = _.map(resolvedRes?.items, (crd) => {
+      const {
+        spec: {
+          group,
+          version,
+          names: { kind, plural, singular },
+        },
+      } = crd;
+      return {
+        apiGroup: group,
+        apiVersion: version,
+        kind,
+        plural,
+        id: singular,
+        label: singular,
+        labelPlural: plural,
+        abbr: kindToAbbr(kind),
+        namespaced: true,
+        crd: true,
+        color: knativeEventingColor.value,
+      };
+    });
+  } catch (err) {
+    // show warning if there is an error fetching the CRDs
+    // eslint-disable-next-line no-console
+    console.warn('Error fetching CRDs for dynamic event sources', err);
```

Comment:
Can we show error modal in place of a console warning? I think that makes more sense.

Reply:
we are passing registered models by default (in case of failure) and don't want user to know if it fails so just added warning.

Reply:
Or we can let the user know there was some issue in fetching the CRDs at the moment and hence we are showing the default.

Reply:
This request is made before the page is loaded. Don't think it's a good idea to show a modal or anything that would halt everything or break. The best we can do is show a warning in the console with some information. Also, We fall back to the default Event sources so it doesn't break anything.